**Q1 Dataset Exploration**

**Perform the following exploratory steps:**

**Display the first 10 rows**

**Display the dataset structure using info()**

**Find the number of rows and columns**

**Identify the data types of each column**

In [ ]:
import pandas as pd

# 1. Load the dataset
# Replace 'Advanced_pandas_sales_dataset.csv' with your actual file path
file_path = 'Advanced_pandas_sales_dataset.csv'
df = pd.read_csv(file_path)

# 2. Display the first 10 rows
print("First 10 rows of the dataset:")
print(df.head(10))

# 3. Display the dataset structure using info()
print("\nDataset Structure:")
df.info()

# 4. Find the number of rows and columns
rows, columns = df.shape
print(f"\nNumber of Rows: {rows}")
print(f"Number of Columns: {columns}")

# 5. Identify the data types of each column
print("\nData Types of each column:")
print(df.dtypes)

First 10 rows of the dataset:
   Order_ID  Order_Date Customer       City Region   Product     Category  \
0      1001  2024-10-01    Priya  Hyderabad   East     Mouse  Accessories   
1      1002  2024-05-03    Sneha    Kolkata   East   Monitor  Electronics   
2      1003  2024-01-26    Anita     Mumbai   West     Phone  Electronics   
3      1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4      1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   
5      1006  2024-07-04    Pooja  Bangalore   West   Printer       Office   
6      1007  2024-06-24      Dev    Chennai   East   Monitor  Electronics   
7      1008  2024-03-18    Kavya    Kolkata   East     Phone  Electronics   
8      1009  2024-10-31    Kavya  Ahmedabad   East     Phone  Electronics   
9      1010  2024-12-25    Pooja  Ahmedabad  North     Phone  Electronics   

   Sales  Quantity  Discount  Profit  
0   1735         7      0.00  307.52  
1    807         2      0.15   49.89  
2   2

**Q2 Sales Filtering Analysis**

**Filter the dataset to display:**

**Orders where Sales > 3500**

**Orders where Region = "East" and Sales > 2500**

**Explain what insights you observe.**

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Orders where Sales > 3500
sales_gt_3500 = df[df['Sales'] > 3500]
print("--- Orders where Sales > 3500 ---")
print(sales_gt_3500.head())
print(f"Number of orders: {len(sales_gt_3500)}")

# 2. Orders where Region = "East" and Sales > 2500
east_high_sales = df[(df['Region'] == 'East') & (df['Sales'] > 2500)]
print("\n--- Orders where Region = 'East' and Sales > 2500 ---")
print(east_high_sales.head())
print(f"Number of orders: {len(east_high_sales)}")

# Summary statistics for insights
print("\nSummary for Sales > 3500:")
print(sales_gt_3500['Category'].value_counts())
print(sales_gt_3500['Profit'].describe())

print("\nSummary for East & Sales > 2500:")
print(east_high_sales['Category'].value_counts())
print(east_high_sales['Profit'].describe())

--- Orders where Sales > 3500 ---
    Order_ID  Order_Date Customer       City Region   Product     Category  \
3       1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4       1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   
7       1008  2024-03-18    Kavya    Kolkata   East     Phone  Electronics   
8       1009  2024-10-31    Kavya  Ahmedabad   East     Phone  Electronics   
10      1011  2024-06-30     Neha      Delhi  South   Printer       Office   

    Sales  Quantity  Discount   Profit  
3    3933         3      0.20   -71.66  
4    3550         2      0.20   146.30  
7    4693         7      0.05   918.49  
8    5071         6      0.10   707.55  
10   5977         6      0.00  1578.32  
Number of orders: 104

--- Orders where Region = 'East' and Sales > 2500 ---
    Order_ID  Order_Date Customer       City Region   Product     Category  \
3       1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
7       1008  2024

**Q3 Customer Purchase Analysis**

**Identify the top 5 customers based on total sales.**

**Hint: Use GroupBy with aggregation**

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Group by Customer and calculate total sales
customer_sales = df.groupby('Customer')['Sales'].sum().reset_index()

# Sort by Sales in descending order and get top 5
top_5_customers = customer_sales.sort_values(by='Sales', ascending=False).head(5)

print("Top 5 Customers based on Total Sales:")
print(top_5_customers)

Top 5 Customers based on Total Sales:
   Customer  Sales
9     Nitin  59820
13      Raj  58416
12    Rahul  52291
3       Dev  48500
4      Isha  46818


**Q4 Regional Sales Performance**

**Calculate the total sales generated by each region.**

**Then identify:**

**Which region generated the highest revenue?**

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Calculate total sales by region
regional_sales = df.groupby('Region')['Sales'].sum().reset_index()

# Find the region with the highest revenue
highest_revenue_region = regional_sales.loc[regional_sales['Sales'].idxmax()]

print("Total Sales by Region:")
print(regional_sales)
print("\nRegion with the highest revenue:")
print(highest_revenue_region)

Total Sales by Region:
  Region   Sales
0   East  232675
1  North  158210
2  South  183108
3   West  170227

Region with the highest revenue:
Region      East
Sales     232675
Name: 0, dtype: object


**Q5 Product Performance Analysis**

**Calculate the total quantity sold for each product.**

**Which product appears to be the most popular?**

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Calculate total quantity sold for each product
product_performance = df.groupby('Product')['Quantity'].sum().reset_index()

# Sort products by Quantity to find the most popular
product_performance_sorted = product_performance.sort_values(by='Quantity', ascending=False)

# Identify the most popular product
most_popular_product = product_performance_sorted.iloc[0]

print("Total Quantity Sold per Product:")
print(product_performance_sorted)

print("\nMost Popular Product:")
print(most_popular_product)

Total Quantity Sold per Product:
      Product  Quantity
3     Monitor       156
5       Phone       151
7      Tablet       150
2      Laptop       140
4       Mouse       122
0  Headphones       115
1    Keyboard       110
6     Printer       108

Most Popular Product:
Product     Monitor
Quantity        156
Name: 3, dtype: object


**Q6 Multi-Level Grouping**

**Group the dataset by: Region and Category**

**Calculate the total sales for each combination.**

**What insights can you derive from this analysis?**

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Multi-level grouping: Region and Category
multi_group_sales = df.groupby(['Region', 'Category'])['Sales'].sum().reset_index()

# Sort for better readability
multi_group_sales = multi_group_sales.sort_values(by=['Region', 'Sales'], ascending=[True, False])

print("Total Sales by Region and Category:")
print(multi_group_sales)

# Pivot for better insight observation
pivot_sales = multi_group_sales.pivot(index='Region', columns='Category', values='Sales')
print("\nPivot Table View:")
print(pivot_sales)

Total Sales by Region and Category:
   Region     Category   Sales
1    East  Electronics  160907
0    East  Accessories   57646
2    East       Office   14122
4   North  Electronics   70023
3   North  Accessories   62052
5   North       Office   26135
7   South  Electronics   86690
6   South  Accessories   80914
8   South       Office   15504
10   West  Electronics   84128
9    West  Accessories   64811
11   West       Office   21288

Pivot Table View:
Category  Accessories  Electronics  Office
Region                                    
East            57646       160907   14122
North           62052        70023   26135
South           80914        86690   15504
West            64811        84128   21288


**Q7 Pivot Table Creation**

**Create a pivot table showing:**

**Rows → Region**

**Columns → Product**

**Values → Total Sales**

**Which region sells the most laptops?**

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Create the pivot table
# Rows -> Region, Columns -> Product, Values -> Sum of Sales
pivot_table = pd.pivot_table(df, values='Sales', index='Region', columns='Product', aggfunc='sum')

print("Pivot Table (Sales by Region and Product):")
print(pivot_table)

# Identify the region with the highest laptop sales
most_laptops_region = pivot_table['Laptop'].idxmax()
print(f"\nThe region that sells the most laptops is: {most_laptops_region}")

Pivot Table (Sales by Region and Product):
Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          23899     10845   55568    31299  22902  39667    14122   34373
North         15450     20538    4540    14922  26064  30535    26135   20026
South         24111     23480   23029    19167  33323  18218    15504   26276
West          17159     38499   10826    18178   9153  27850    21288   27274

The region that sells the most laptops is: East


**Q8 Profitability Analysis**

**Calculate the average profit for each product category.**

**Which category is most profitable?**

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Calculate the average profit for each category
category_profitability = df.groupby('Category')['Profit'].mean().reset_index()

# Sort the results by Profit in descending order
category_profitability = category_profitability.sort_values(by='Profit', ascending=False)

# Identify the most profitable category
most_profitable_category = category_profitability.iloc[0]

print("Average Profit per Category:")
print(category_profitability)

print(f"\nThe most profitable category (by average profit) is {most_profitable_category['Category']} "
      f"with an average profit of {most_profitable_category['Profit']:.2f}")

Average Profit per Category:
      Category      Profit
2       Office  471.938462
0  Accessories  410.603086
1  Electronics  396.462114

The most profitable category (by average profit) is Office with an average profit of 471.94


**Q9 Discount Impact Study**

**Analyze the relationship between discount and profit.**

**Tasks:**

**Calculate the average profit for each discount level**

**Identify whether higher discounts lead to lower profit**

In [ ]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# Calculate the average profit for each discount level
discount_impact = df.groupby('Discount')['Profit'].mean().reset_index()

# Sort by Discount to observe the trend
discount_impact = discount_impact.sort_values(by='Discount')

print("Average Profit per Discount Level:")
print(discount_impact)

# Calculate correlation to quantify the relationship
correlation = df['Discount'].corr(df['Profit'])
print(f"\nCorrelation between Discount and Profit: {correlation:.4f}")



Average Profit per Discount Level:
   Discount      Profit
0      0.00  751.210588
1      0.05  531.368039
2      0.10  346.460513
3      0.15  237.191400
4      0.20   90.026923

Correlation between Discount and Profit: -0.6637


Conclusion: Yes, higher discounts significantly lead to lower profits in this dataset. While discounts might drive volume (sales quantity), they appear to be cutting deeply into the margins.

**Q10 End-to-End Data Wrangling Task**

**Perform the following workflow:**

**Filter the dataset where Sales > 3000**

**Group data by Region and Product**

**Calculate total sales**

**Reshape the result into a pivot table**

**Explain the business insights obtained from this analysis**

In [12]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

# 1. Filter the dataset where Sales > 3000
filtered_df = df[df['Sales'] > 3000]

# 2 & 3. Group data by Region and Product and calculate total sales
grouped_sales = filtered_df.groupby(['Region', 'Product'])['Sales'].sum().reset_index()

# 4. Reshape the result into a pivot table
# Filling missing values (NaN) with 0 for better readability
pivot_table = grouped_sales.pivot(index='Region', columns='Product', values='Sales').fillna(0)

print("Final Pivot Table (Sales > 3000):")
print(pivot_table)

Final Pivot Table (Sales > 3000):
Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          14469      9642   44096    16420  19000  29500     3696   29229
North          5326     15653    4540     5087  22209  24670    17185    3392
South         10026     19249   17950    14921  26928  18218    14143   19613
West          13898     33380    5922    14354   3511  15653    12648   17185


Business Insights ObtainedHigh-Value Dominance in the East: The East region stands out significantly for high-value Laptop sales ($44,096$). This suggests that the East is a primary market for premium hardware or corporate bulk purchases.Product Specificity by Region: * The West region has a very high concentration of high-value Keyboard sales ($33,380$), which is nearly double that of any other region for that specific product.The North region shows lower totals for high-value Laptops and Tablets, indicating that while these products are sold there, they may usually fall under the $3000$ price threshold.High-Value Sales Volume: The East region has the highest number of high-value orders ($37$), followed by the South ($31$). This identifies these two regions as the most "premium" markets in the dataset.Leading Premium Category: Phones and Keyboards are the most consistent drivers of high-revenue orders across the board. If the company wants to increase revenue, focusing on marketing premium Phone models in the North and West could be a strategic growth opportunity.